rr_evgp/src/UniformCostSearch/UniformCostSearch.cpp

```diff
@@ -0,0 +1,126 @@
+#include <rr_evgp/UniformCostSearch.h>
+#include <algorithm>
+#include <functional>
+#include <opencv2/core/mat.hpp>
+#include <opencv2/core/types.hpp>
+#include <queue>
+#include <vector>
+
+UniformCostSearch::UniformCostSearch(cv::Mat obstacleGrid, cv::Mat distanceGrid, cv::Point startPt, cv::Point goalPt) {
+    obstacleGrid_ = obstacleGrid;
+    distanceGrid_ = distanceGrid;
+    startPoint_ = startPt;
+    goalPoint_ = goalPt;
+}
+
+UniformCostSearch::State UniformCostSearch::getStartState() {
+    UniformCostSearch::State s;
+    s.pt = startPoint_;
+    s.cost = 0.0;
+    return s;
+}
+
+bool UniformCostSearch::isGoalState(UniformCostSearch::State s) {
+    if (s.pt == goalPoint_) {
+        return true;
+    }
+    return false;
+}
+
+float UniformCostSearch::pointToCost(cv::Point pt) {
+    return distanceGrid_.at<float>(pt);  // DistanceTransform outputs CV_32F
+}
+
+bool UniformCostSearch::isValidPoint(cv::Point pt) {
+    cv::Rect rect(cv::Point(), distanceGrid_.size());
+    return rect.contains(pt) && obstacleGrid_.at<uchar>(pt) == UniformCostSearch::FREE_SPACE;
+}
+
+std::vector<UniformCostSearch::State> UniformCostSearch::getSuccessors(UniformCostSearch::State s) {
+    // returns the 8 point grid around a center point
+    std::vector<UniformCostSearch::State> successors;
+    for (int i = -1; i < 2; i++) {
+        for (int j = -1; j < 2; j++) {
+            if (i != 0 || j != 0) {  // not center position
+                cv::Point pt(s.pt.x + i, s.pt.y + j);
+                if (this->isValidPoint(pt)) {
+                    UniformCostSearch::State newState;
+                    newState.pt = pt;
+                    newState.cost = this->pointToCost(pt);
+                    successors.push_back(newState);
+                }
+            }
+        }
+    }
+    return successors;
+}
+
+std::vector<cv::Point> UniformCostSearch::search() {
+    std::priority_queue<UniformCostSearch::State, std::vector<UniformCostSearch::State>,
+                        std::greater<UniformCostSearch::State>>
+          pq;  // minimum priority queue
+    cv::Mat visited(obstacleGrid_.size(), CV_8UC1, cv::Scalar(0));
+
+    pq.push(this->getStartState());
+    // uniform cost search
+    while (!pq.empty()) {
+        UniformCostSearch::State e = pq.top();
+        if (visited.at<uchar>(e.pt) == 0) {
+            if (this->isGoalState(e)) {
+                return e.path;
+            }
+            std::vector<UniformCostSearch::State> successors = this->getSuccessors(e);
```

Comment:
Could just return `vector<cv::Point>` and just get state cost when adding to parent state cost. Both work fine